sweeper/sweeper.go

```diff
@@ -0,0 +1,191 @@
+package sweeper
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/skynetlabs/pinner/database"
+	"github.com/skynetlabs/pinner/skyd"
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/build"
+	"gitlab.com/SkynetLabs/skyd/skymodules"
+)
+
+type (
+	// Schedule defines how often, if at all, we sweep this server automatically.
+	Schedule struct {
+		Period   time.Duration
+		cancelCh chan struct{}
+	}
+	// Status represents the status of a sweep.
+	Status struct {
+		InProgress bool
+		Error      error
+		StartTime  time.Time
+		EndTime    time.Time
+	}
+	// Sweeper takes care of sweeping the files pinned by the local skyd server
+	// and marks them as pinned by the local server.
+	Sweeper struct {
+		staticDB         *database.DB
+		staticSkydClient skyd.Client
+		staticServerName string
+
+		schedule   *Schedule
+		scheduleMu sync.Mutex
+
+		status   Status
+		statusMu sync.Mutex
+	}
+)
+
+// New returns a new Sweeper.
+func New(db *database.DB, skydc skyd.Client, serverName string) *Sweeper {
+	return &Sweeper{
+		staticDB:         db,
+		staticSkydClient: skydc,
+		staticServerName: serverName,
+	}
+}
+
+// Schedule a scan to run on each period.
+func (s *Sweeper) Schedule(period time.Duration) {
+	s.scheduleMu.Lock()
+	defer s.scheduleMu.Unlock()
+	if s.schedule != nil {
+		close(s.schedule.cancelCh)
+	}
+	s.schedule = &Schedule{
+		Period:   period,
+		cancelCh: make(chan struct{}),
+	}
+	go func() {
+		t := time.NewTicker(s.schedule.Period)
+		for {
+			select {
+			case <-t.C:
+				s.Sweep()
+			case <-s.schedule.cancelCh:
+				return
+			}
+		}
+	}()
+}
+
+// Scheduled returns a copy of the current Sweeper schedule.
+func (s *Sweeper) Scheduled() Schedule {
```

Comment:
`Schedule` or `ActiveSchedule`. `Scheduled` sounds like the result would be a `bool`.

Reply:
Also is this unused?

Reply:
> `Scheduled` sounds like the result would be a `bool`.

This is so counter-intuitive to me. :( In my mind boolean functions are `IsScheduled` and such. And "schedule" is both a noun and a verb, that's why `Schedule(period)` sounded as a good constructor name.

> Also is this unused?

Right, maybe we don't need it, yet. I'll remove it for the moment.